powerauth-java-cmd-lib/src/main/java/io/getlime/security/powerauth/lib/cmd/util/CounterUtil.java

```diff
@@ -40,8 +41,8 @@
      */
     public static byte[] getCtrData(BaseStepModel model, StepLogger stepLogger) {
         byte[] ctrData = new byte[16];
-        long counter = (long) model.getResultStatusObject().get("counter");
-        int version = (int) model.getResultStatusObject().get("version");
+        long counter = new BigDecimal(model.getResultStatusObject().get("counter").toString()).longValue();
```

Comment:
This seems a bit overengineered: `TooManyConversionException`. :)

I would probably favor something like:

```java
long counter = safeLong(model.getResultStatusObject().get("counter"));

/// ...

private long safeLong(Object o) {
    return o instanceof Long ? (long) o : 0;
}

```

Reply:
Ok, why not.

Reply:
I introduced safe handling of `long`, `int` and `String` values to avoid dangerous casts in the code. The ObjectMapper is configured to return `Long` type for deserialized numbers instead of `Integer` so that the same type is used in memory and during deserialization. Previously the type was either `Long` (in memory counter) or `Integer` (counter read from file) which lead to class cast exceptions.